File: 10Feb2024.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        def f(s1:str,i:int,j:int) :
            if(i>=j) :
                return 1
            if(dp[i][j]!=-1) :
                return dp[i][j]
            ans=0;
            if(s1[i]==s1[j]) :
                ans+=f(s,i+1,j-1)
            dp[i][j]=ans
            return ans
        n = len(s)
        dp = [[-1]*n for _ in range(n)]
        ans=0
        for i in range(n) :
            for j in range(i,n) :
                ans+=f(s,i,j)
        return ans

File: 10Feb2024.java
class Solution {
    public int countSubstrings(String s) {
        Integer[][] dp = new Integer[s.length()][s.length()];
        int ans = 0;
        for (int i = 0; i < s.length(); i++) {
            for (int j = i; j < s.length(); j++) {
                ans += f(s, i, j, dp);
            }
        }
        return ans;
    }

    public int f(String s, int i, int j, Integer[][] dp) {
        // System.out.println(i+" "+j);
        if (i >= j)
            return dp[i][j] = 1;
        if (dp[i][j] != null)
            return dp[i][j];
        int ans = 0;
        if (s.charAt(i) == s.charAt(j)) {
            ans += f(s, i + 1, j - 1, dp);
        }
        return dp[i][j] = ans;
    }
}